cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java

```diff
@@ -633,42 +638,81 @@ public void run() throws Exception {
   /**
    * Destroy a program. The destroy() method is executed with the context class loader combined from the
    * program class loader and the system class loader. If the transaction control is implicit, then this code
-   * is wrapped into a transaction, possibly with retry on conflict.
+   * is wrapped into a transaction, possibly with retry on conflict. This method never throw exception. Any error
+   * raised during the destroy lifecycle call will just be logged.
    *
    * @param program the program to be destroyed
-   * @param programContext the program context
    * @param txControl the transaction control
    * @param retryOnConflict if true, transactional execution will be retried on conflict
-   * @param <T> the type of the program context
    */
-  public <T extends AbstractContext> void destroyProgram(final ProgramLifecycle<? super T> program,
-                                                         final T programContext, TransactionControl txControl,
-                                                         boolean retryOnConflict)
-    throws TransactionFailureException {
-    if (TransactionControl.IMPLICIT == txControl) {
-      programContext.execute(new TxRunnable() {
-        @Override
-        public void run(DatasetContext context) throws Exception {
-          program.destroy();
-        }
-      }, retryOnConflict);
-    } else {
-      programContext.executeUnchecked(new Runnable() {
-        @Override
-        public void run() {
-          program.destroy();
+  public void destroyProgram(final ProgramLifecycle<?> program, TransactionControl txControl, boolean retryOnConflict) {
+    try {
+      try {
+        if (TransactionControl.IMPLICIT == txControl) {
+          execute(new TxRunnable() {
+            @Override
+            public void run(DatasetContext context) throws Exception {
+              program.destroy();
+            }
+          }, retryOnConflict);
+        } else {
+          executeChecked(new ThrowingRunnable() {
+            @Override
+            public void run() throws Exception {
+              program.destroy();
+            }
+          });
         }
-      });
+      } catch (TransactionConflictException e) {
+        // For Tx conflict, use it as is.
+        throw e;
+      } catch (TransactionFailureException | UncheckedExecutionException e) {
+        // For tx failure (implicit case) or unchecked execution exception (explicit with CDAP wrapper case),
+        // throw the cause for logging.
+        throw e.getCause() == null ? e : e.getCause();
```

Comment:
seems like all this logic doesn't matter, as it just gets caught as a Throwable below, and we don't do anything with it. Should we at least log the throwable below?

Reply:
Oh, missing the last argument. Will add it.